### PR TITLE
Integrate check-bib-for-predatory

### PR DESCRIPTION
**Pull request by milescajus**
Fixes https://github.com/koppor/jabref/issues/348 by creating an equivalent functionality in Java.

The implementation attempts to match the architecture of the `JournalInAbbreviationListChecker`.

This draft pull request will be updated as the implementation is completed.

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

File: src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+    private static void update() {
+        PREDATORY_SOURCES   .forEach(PredatoryJournalLoader::crawl);            // populates linkElements (and predatoryJournals if CSV)
+        linkElements        .forEach(PredatoryJournalLoader::clean);            // adds cleaned HTML to predatoryJournals
+
+        LOGGER.info("UPDATED PREDATORY JOURNAL LIST");
+    }
+
+    private static void crawl(PJSource source) {
+        var uri     = URI.create(source.URL);
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java:77]: You can use our class UrlDownload for downloading/getting data. That already comes with exception handling etc

File: src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            else                                    { handleHTML(source.ELEMENT_REGEX, response.body()); }
+        }
+        catch (IOException ex)          { logException(ex); }
+        catch (InterruptedException ex) { logException(ex); }
+    }
+
+    private static void handleCSV(String body) {
+        var csvSplit = Pattern.compile("(\"[^\"]*\"|[^,]+)");
+
+        for (String line : body.split("\n")) {                                  // TODO: skip header
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java:94]: You can use the CSV library we already have included
 https://github.com/JabRef/jabref/blob/9d3bcf3d98aa8eedfdf0ca2ec9790baa31291f83/src/main/java/org/jabref/logic/journals/AbbreviationParser.java#L57

**Comment by Siedlerchr** [comment]
thanks for you interest in this and your contribution. Some general thing: Your code style looks a bit odd. https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace.html#using-jabrefs-code-style

**Comment by milescajus** [comment]
Thanks for the feedback! I will address the mentioned aspects.

File: src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java (hunk 2 above)
Comment (milescajus) [reply src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java:94]: Ah good, I didn't see that there was one included. Will switch to that.

**Comment by milescajus** [comment]
The deployment check for macOS is failing with the following error message:
`Error: At least one of p12-filepath or p12-file-base64 must be provided`

Is there something I need to do for this to pass?

**Comment by Siedlerchr** [comment]
Hi you can ignore the failing mac test. I does not work on forks because it require some github secrets for signing. We are working on trying to remove this run then for forks

File: src/main/java/org/jabref/logic/integrity/IntegrityCheck.java [hunk 3]
```diff
@@ -42,11 +44,13 @@ public IntegrityCheck(BibDatabaseContext bibDatabaseContext,
         if (bibDatabaseContext.isBiblatexMode()) {
             entryCheckers.addAll(List.of(
                     new JournalInAbbreviationListChecker(StandardField.JOURNALTITLE, journalAbbreviationRepository),
+                    new PredatoryJournalChecker(predatoryJournalRepository, StandardField.JOURNALTITLE, StandardField.PUBLISHER, StandardField.BOOKTITLE),
```
Comment (koppor) [inline src/main/java/org/jabref/logic/integrity/IntegrityCheck.java:47]:
```suggestion
                    new PredatoryJournalChecker(predatoryJournalRepository, List.of(StandardField.JOURNALTITLE, StandardField.PUBLISHER, StandardField.BOOKTITLE)),
```

File: src/main/java/org/jabref/logic/integrity/PredatoryJournalChecker.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import org.jabref.logic.l10n.Localization;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.entry.field.Field;
+
+public class PredatoryJournalChecker implements EntryChecker {
+
+    private final List<Field> fieldsToCheck;
+    private final PredatoryJournalRepository predatoryJournalRepository;
+
+    public PredatoryJournalChecker(PredatoryJournalRepository predatoryJournalRepository, Field... fieldsToCheck) {
```
Comment (koppor) [inline src/main/java/org/jabref/logic/integrity/PredatoryJournalChecker.java:21]: JabRef's "style" is to pass Lists. With `List.of(...)` this should be "easy" for the caller.

Then you can also remove the for loop from lines 24 to 27

File: src/main/java/org/jabref/logic/integrity/PredatoryJournalChecker.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        if (fields.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        for (Map.Entry<Field, String> field : fields.entrySet()) {
+            if (predatoryJournalRepository.isKnownName(field.getValue(), 0.95)) {
+                results.add(new IntegrityMessage(Localization.lang("match found in predatory journal list"), entry, field.getKey()));
+            }
+        }
+
```
Comment (koppor) [inline src/main/java/org/jabref/logic/integrity/PredatoryJournalChecker.java:51]: We came up with a shorter code.

We also added the found journal to the output message.

```suggestion
        List<Map.Entry<Field, String>> entries = entry.getFieldMap().entrySet().stream()
                                                         .filter(field -> fieldNames.contains(field.getKey().getName()))
                                                         .filter(field -> predatoryJournalRepository.isKnownName(field.getValue(), 0.95))
                                                         .toList();
        for (var mapEntry : entries) {
            results.add(new IntegrityMessage(Localization.lang("Match found in predatory journal %0", mapEntry.getValue()), entry, mapEntry.getKey()));
        }
```

File: src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+import kong.unirest.UnirestException;
+import org.apache.commons.csv.CSVFormat;
+import org.apache.commons.csv.CSVParser;
+import org.apache.commons.csv.CSVRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PredatoryJournalLoader {
+    private static class PJSource {
```
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java:24]: Isn't this a record?

File: src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+        URL URL = null;
+        final String ELEMENT_REGEX;
```
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java:26]: Variables, fields, ... should be lower case. Please change to lower case. (also at the constructor below)

File: src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+            Files.copy(JournalAbbreviationRepository.class.getResourceAsStream("/journals/predatoryJournalList.mv"), tempJournalList);
+            repository = new PredatoryJournalRepository(tempJournalList);
+            tempDir.toFile().deleteOnExit();
+            tempJournalList.toFile().deleteOnExit();
+        } catch (IOException e) {
+            LOGGER.error("Error while copying journal list", e);
+            return null;
+        }
+
+        // Update from external sources
```
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java:72]: Remove this comment - the method name already states that

File: src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    );
+    private static PredatoryJournalRepository repository;
+    private static List<String> linkElements = new ArrayList<>();
+
+    public static PredatoryJournalRepository loadRepository(boolean doUpdate) {
+        // Initialize repository
+        try {
+            Path tempDir = Files.createTempDirectory("jabref-journal");
+            Path tempJournalList = tempDir.resolve("predatoryJournalList.mv");
+            Files.copy(JournalAbbreviationRepository.class.getResourceAsStream("/journals/predatoryJournalList.mv"), tempJournalList);
```
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java:63]: Please comment why a copy is needed of this file. Can't it be updated in place?

File: src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+        if (doUpdate) {
+            update();
+        }
+
+        return repository;
+    }
+
+    private static void update() {
+        PREDATORY_SOURCES.forEach(PredatoryJournalLoader::crawl);       // populates linkElements (and predatoryJournals if CSV)
+        linkElements.forEach(PredatoryJournalLoader::clean);            // adds cleaned HTML to predatoryJournals
```
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java:82]: Please put one line comments above the statements

File: src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+        LOGGER.info("UPDATED PREDATORY JOURNAL LIST");
```
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java:84]: Please normal sentence case at log meesages.

File: src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+        if (m_name.find() && m_url.find()) {
+            repository.addToPredatoryJournals(m_name.group(), m_abbr.find() ? m_abbr.group() : "", m_url.group());
+        }
+    }
+
+    private static void logException(Exception ex) {
+        if (LOGGER.isErrorEnabled()) {
+            LOGGER.error(ex.getMessage(), ex);
+        }
+    }
```
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java:139]: Please inline. `LOGGER.error` already checks whether it is enabled.

Loggers have to be

    LOGGER.error("text", ex);

`text` should be something in the context of the current method, not the ex.getMesasge

File: src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+            } else if (source.URL.getPath().contains(".csv")) {
+                handleCSV(new InputStreamReader(download.asInputStream()));
+            } else {
+                handleHTML(source.ELEMENT_REGEX, download.asString());
+            }
+        } catch (IOException ex) {
+            logException(ex);
+        } catch (UnirestException ex) {
+            logException(ex);
+        }
```
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java:102]:
```suggestion
        } catch (Exception ex) {
            LOGGER.error("Could not crawl source {}", source.URL, ex);
        }
```

File: src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+
+        while (matcher.find()) {
+            linkElements.add(matcher.group());
+        }
+    }
+
+    private static void clean(String item) {
+        var m_name = Pattern.compile("(?<=\">).*?(?=<)").matcher(item);
+        var m_url = Pattern.compile("http.*?(?=\")").matcher(item);
+        var m_abbr = Pattern.compile("(?<=\\()[^\s]*(?=\\))").matcher(item);
```
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java:126]: Please work with `Matcher` to have the Patterns pre-comiled on class startup (and not on each run of this method)

File: src/main/java/org/jabref/logic/journals/PredatoryJournalRepository.java [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+            return true;
+        }
+
+        var matches = predatoryJournals.keySet().stream()
+                                                .filter(key -> cosineSimilarity(key, journal) > similarityScore)
+                                                .collect(Collectors.toList());
+
+        if (LOGGER.isInfoEnabled()) {
+            LOGGER.info("matches: " + String.join(", ", matches));
+        }
```
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/PredatoryJournalRepository.java:49]:
```suggestion
        LOGGER.info("matches:{}", String.join(", ", matches));
```

File: src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+                    "<li>.*?</li>"),
+        new PJSource("https://beallslist.net/standalone-journals/",
+                    "<li>.*?</li>"),
+        new PJSource("https://beallslist.net/hijacked-journals/",
+                    "<tr>.*?</tr>")
+    );
+    private static PredatoryJournalRepository repository;
+    private static List<String> linkElements = new ArrayList<>();
+
+    public static PredatoryJournalRepository loadRepository(boolean doUpdate) {
```
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java:58]: Do not mix update with loading. Please remove the boolean parameter and never (!) update the list.

File: src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+
+        // Update from external sources
+        if (doUpdate) {
+            update();
+        }
+
+        return repository;
+    }
+
+    private static void update() {
```
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java:80]: Make this a public method. Can this be non-static?

File: src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+        PREDATORY_SOURCES.forEach(PredatoryJournalLoader::crawl);       // populates linkElements (and predatoryJournals if CSV)
+        linkElements.forEach(PredatoryJournalLoader::clean);            // adds cleaned HTML to predatoryJournals
+
+        LOGGER.info("UPDATED PREDATORY JOURNAL LIST");
+    }
+
+    private static void crawl(PJSource source) {
```
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/PredatoryJournalLoader.java:87]: This should be non-static

**Review by koppor — CHANGES_REQUESTED** [review]
In general, looks good. Some code comments.

The update method should be made a JabRef CLI option - not an automatic update.

Reason: the lists update very seldom and we should avoid huge network traffic.

Only JabRef maintainers (or motivated users) should execute the CLI to update the .mv store.

And when update the .mv store, the internal one of JabRef should be updated and not some temporary one.

**Comment by koppor** [comment]
@milescajus Happy new year 🎉. We had a busy time here end of the year and thus we needed longer for a code review. Would you be willing to continue on the PR - or should we take over?

**Comment by milescajus** [comment]
Happy New Year! I'm more than happy to keep working on the PR, addressing the changes you've requested.

**Comment by Siedlerchr** [comment]
closing this now due to inactivity

**Pull request closed without merging**